05-mvc-with-rails/rails-sessions-and-auth-with-devise/readme.md

```diff
@@ -0,0 +1,319 @@
+- A session is just a place to store data during one request that you can read during later requests.
+- A session is a hash containing key value pairs that provide state in your application
+
+Let's hop into some code so we can see sessions in action.
+
+The code we'll be working with today can be found [here](https://github.com/ga-dc/tunr_rails_sessions_devise) This is a version of the tunr rails application we've been building on this week. It contains everything that we've learned up until now.
+
+Make sure you fork/clone this repo if you would like to follow along. It is not too important to follow along for this portion of this lesson. I encourage you to just watch and use the following code to practice it yourself later.
+
+## Set and Use session key-value pairs
+
+In order to understand sessions, we're going to add some code to our existing artists controller. In `app/controllers/artists_controller.rb`:
+
+```ruby
+# this action will set `test: "this string was set on the testing_session action"` as a key value pair in the session
+def testing_session
+  session[:test] = "this string was set on the testing_session action"
+end
+
+# also add @test = session[:test] to index action
+def index
+  @test = session[:test]
+  @artists = Artist.all
+end
+```
+
+Let's update our `app/views/artists/index.html.erb` to include the use of our new instance variable `@test`. Place the following code somewhere in that file:
+
+```html
+<h1>The value of @test is: <%= @test %></h1>
+```
+
+Now because we created this action, let's update our `config/routes.rb` to listen for a request to that action:
+
+```ruby
+get '/artists/testing_session', to: 'artists#testing_session'
+```
+
+Finally we need to create a view for this action. Lets create that now `$ touch views/artists/testing_session.html.erb`. Let's make this view simple and just put in the following content.
+
+```html
+<h1>A session key value pair is being set</h1>
+```
+
+Let's first navigate to our artists index route at `http://localhost:3000/artists`.
+
+We can see that there is nothing after `The value of @test is: `
+
+Now let's navigate to `http://localhost:3000/artists/testing_session`
+
+Great that's nice, our view was generated, but more importantly we had a key-value pair get set in this request because of the `testing_session` action in our artists controller:
+
+```ruby
+def testing_session
+  session[:test] = "this was set on the testing_session action"
+end
+```
+
+How do we verify this session key-value pair was set? Let's navigate back to artists index path.
+
+Now we can see `The value of @test is: this was set on the testing_session action`
+
+If we open a new incognito window and navigate to `http://localhost:3000/artists`, we'll see the session isn't set in that browser because it's a brand new session.
+
+When we clear our cookies/session in our browser we can see in our index route that `@test` becomes nil again.
+
```

Comment:
I think it's really valuable for students to explore cookies too, because sessions can't exist without them. I think the Cookie Monster and Sessions sections here are good at introducing the two in such a way that they don't get mixed up:

https://github.com/ga-dc/curriculum/tree/master/05-mvc-with-rails/rails-sessions-and-auth#cookie-monster
https://github.com/ga-dc/curriculum/tree/master/05-mvc-with-rails/rails-sessions-and-auth#sessions